src/Std/Time/Format.lean

```diff
@@ -0,0 +1,505 @@
+  let format : Except String (Format .any) := Format.spec format
+  match format with
+  | .error err => s!"error: {err}"
+  | .ok res =>
+    let res := res.formatGeneric fun
+      | .YYYY | .YY => some date.year
+      | .MMMM | .MMM | .MM | .M => some date.month
+      | .DD | .D | .d => some date.day
+      | .EEEE | .EEE => some date.date.weekday
+      | .HH | .H => some date.time.hour
+      | .hh | .h => some date.time.hour
+      | .aa | .AA => some (if date.time.hour.snd.val > 12 then HourMarker.pm else HourMarker.am)
+      | .mm | .m => some date.time.minute
+      | .sssssssss => some date.time.nano
+      | .sss => some date.time.nano.toMillisecond
+      | .ss | .s => some date.time.second
+      | _ => none
+    match res with
+    | some res => res
+    | none => "invalid time"
+
+/--
+Parses a `String` in the `AscTime` format and returns a `PlainDateTime` object in the GMT time zone.
+-/
+def fromAscTimeString (input : String) : Except String PlainDateTime :=
+  Formats.ascTime.parse input
+  |>.map DateTime.toPlainDateTime
+
+/--
+Formats a `PlainDateTime` value into an AscTime format string.
+-/
+def toAscTimeString (pdt : PlainDateTime) : String :=
+  Formats.ascTime.format (DateTime.ofPlainDateTime pdt .UTC)
+
+/--
+Parses a `String` in the `LongDateFormat` and returns a `PlainDateTime` object in the GMT time zone.
+-/
+def fromLongDateFormatString (input : String) : Except String PlainDateTime :=
+  Formats.longDateFormat.parse input
+  |>.map DateTime.toPlainDateTime
+
+/--
+Formats a `PlainDateTime` value into a LongDateFormat string.
+-/
+def toLongDateFormatString (pdt : PlainDateTime) : String :=
+  Formats.longDateFormat.format (DateTime.ofPlainDateTime pdt .UTC)
+
+/--
+Parses a `String` in the `DateTime` format and returns a `PlainDateTime`.
+-/
+def fromDateTimeString (input : String) : Except String PlainDateTime :=
+  Formats.dateTime24Hour.parse input
+  |>.map DateTime.toPlainDateTime
+
+/--
+Formats a `PlainDateTime` value into a `DateTime` format string.
+-/
+def toDateTimeString (pdt : PlainDateTime) : String :=
+  Formats.dateTime24Hour.format (DateTime.ofPlainDateTime pdt .UTC)
+
+/--
+Parses a `String` in the `DateTime` format and returns a `PlainDateTime`.
+-/
+def fromLeanDateTimeString (input : String) : Except String PlainDateTime :=
+  Formats.leanDateTime24Hour.parse input
+  |>.map DateTime.toPlainDateTime
+
+/--
+Formats a `PlainDateTime` value into a `DateTime` format string.
+-/
+def toLeanDateTimeString (pdt : PlainDateTime) : String :=
+  Formats.leanDateTime24Hour.format (DateTime.ofPlainDateTime pdt .UTC)
+
+/--
+Parses a `String` in the `AscTime` or `LongDate` format and returns a `PlainDateTime`.
+-/
+def parse (date : String) : Except String PlainDateTime :=
+  fromAscTimeString date
+  <|> fromLongDateFormatString date
+  <|> fromDateTimeString date
+
+instance : ToString PlainDateTime where
+  toString := toLeanDateTimeString
+
+instance : Repr PlainDateTime where
```

Comment:
This currently return something like `2024-08-30:08:56:43:276801081`, which is almost right, but the idea is that it actually round-trips, so it should return `date% 2024-08-30:08:56:43:276801081`.